### PR TITLE
[ENH] Adds `get_stats` method to DAG class

### LINKED ISSUE #2919: [ENH] Add `get_stats` method to `DAG` class
The DAG class represents the graphical structure underlying all of the BN classes. It would be nice to have a `get_stats` method implemented that returns some summary statistics on the structure of the DAG. The following stats can be included:

1. Number of nodes
2. Number of root nodes
3. Number of leaf nodes
4. Number of edges
5. Edge density
6. No. of connected components
7. No of v-structures
8. Average number of parents (average markov blanket size)
9. Maximum number of parents of a node (max in-degree)

### PR DESCRIPTION
**Pull request by Anushka-2406**
# The following checklist is mandatory.

Your PR will be closed if you remove the checklist or do not answer the questions to a satisfactory level. Use of LLM is **strictly forbidden** for any part of this checklist (even for improving language).

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

Please answer the following questions:

- Did you use an LLM for any assistance? Please describe how and what you used it for?
Yes, to check how DAG class is using with networkx and what functions can be directly used to compute the values.

- What steps have you taken to verify that the changes correctly address the issue? And what edge cases have you considered?
Ran the code first on colab, LLM codes iterated over `self.in_degree(node)` multiple times, so I stored it as a variable. Similarly did for nodes and out_degree.
For edge cases : I tried passing single node, empty DAG and disconnected graph.

- Has the LLM added try-except blocks? They will need to be removed; any error handling must be explicit.
No try-except blocks.

- Have you used LLM for generating tests? They need to be compressed into a smaller number of tests without reducing coverage.
Only one test is created which checks all the newly created properties. It also checks the edge cases (disconnected graph and single node). 

### Issue number(s) that this pull request fixes
- Fixes #2919 

### List of changes to the codebase in this pull request
- Added get_stats method in DAG class.
- Created a test for checking get_stats method in `test_DAG.py`.

### CODE REVIEW

File: pgmpy/base/DAG.py [hunk 1]
```diff
@@ -1782,3 +1782,51 @@ def __hash__(self):
...
+        dict : A dictionary with the following keys:
+            - n_nodes: int - Number of nodes in the DAG.
+            - n_edges: int - Number of edges in the DAG.
+            - n_root_nodes: int - Number of root nodes (nodes with no parents).
+            - n_leaf_nodes: int - Number of leaf nodes (nodes with no children).
+            - edge_density: float - Ratio of edges to maximum possible edges.
+            - n_connected_components: int - Number of weakly connected components.
+            - n_v_structures: int - Number of v-structures (immoralities) in the DAG.
+            - avg_n_parents: float - Average number of parents per node.
+            - max_n_parents: int - Maximum number of parents of any single node (max in-degree).
```
Comment (DARHWOLF) [inline pgmpy/base/DAG.py]: Can you format this according to the style guide here 
https://numpydoc.readthedocs.io/en/latest/format.html#returns

**Review by DARHWOLF — CHANGES_REQUESTED** [review]
Nice work!
- I would suggest including a test with a slightly more number of nodes and multiple v-structures since currently the test contains only one. 

- Also, would it be a good idea to include something opposite of the current v-structure -- like  `X<-- U -->Y` ? This could tell us the number of confounding components in the DAG

File: pgmpy/tests/test_base/test_DAG.py [hunk 2]
```diff
@@ -718,6 +716,22 @@ def test_latents_without_role(self):
         self.assertEqual(self.dag1.latents, set())
         self.assertEqual(set(self.dag1.get_role("latents")), set())
 
+    def test_get_stats(self):
```
Comment (ankurankan) [inline pgmpy/tests/test_base/test_DAG.py:721]: Can you also test with a real model from the bnlearn repository: https://www.bnlearn.com/bnrepository/? They have some of the stats on the page and we can use that to compare.

File: pgmpy/base/DAG.py [hunk 3]
```diff
@@ -1782,3 +1782,51 @@ def __hash__(self):
...
+        out_degrees = [self.out_degree(node) for node in nodes]
+
+        n_v_structures = sum(len(pairs) for pairs in self.get_immoralities().values())
+
+        return {
+            "n_nodes": n,
+            "n_edges": e,
+            "n_root_nodes": sum(d == 0 for d in in_degrees),
+            "n_leaf_nodes": sum(d == 0 for d in out_degrees),
+            "edge_density": e / (n * (n - 1)) if n > 1 else 0,
```
Comment (ankurankan) [inline pgmpy/base/DAG.py]: This is incorrect. The max number of edges in a DAG is `n * (n-1) / 2`

File: pgmpy/base/DAG.py [hunk 4]
```diff
@@ -1782,3 +1782,51 @@ def __hash__(self):
...
+        >>> from pgmpy.base import DAG
+        >>> dag = DAG([("D", "G"), ("I", "G"), ("G", "L"), ("I", "S")])
+        >>> stats = dag.get_stats()
+        >>> stats["n_nodes"]
+        5
+        >>> stats["n_v_structures"]
+        1
+        """
+        n = self.number_of_nodes()
+        e = self.number_of_edges()
```
Comment (ankurankan) [inline pgmpy/base/DAG.py]: Can we use better names instead of `n` and `e`?

File: pgmpy/base/DAG.py [hunk 5]
```diff
@@ -1782,3 +1782,51 @@ def __hash__(self):
...
+
+        nodes = list(self.nodes())
+        in_degrees = [self.in_degree(node) for node in nodes]
```
Comment (ankurankan) [inline pgmpy/base/DAG.py]: Calling `in_degree` without an argument returns a dict with in degree for all nodes. It would be simpler than doing the list comprehension.

**Review by ankurankan — CHANGES_REQUESTED** [review]

File: pgmpy/tests/test_base/test_DAG.py (hunk 2 above)
Comment (Anushka-2406) [reply pgmpy/tests/test_base/test_DAG.py:721]: Ok I will add that!

File: pgmpy/base/DAG.py (hunk 3 above)
Comment (Anushka-2406) [reply pgmpy/base/DAG.py]: Sorry about that, I just applied permutation for the two nodes and forgot to put the acyclic condition. Thanks for pointing it out, I'll correct it.

**Comment by Anushka-2406** [comment]
@ankurankan Should I also add `n_confounding_structure`? I think for a node with out degree k, the no. of confounding structures will be kC2.

**Comment by ankurankan** [comment]
@Anushka-2406 @DARHWOLF I wasn't thinking of causal statistics for this issue, as that would require the roles `exposures` and `outcomes` to be defined. @Anushka-2406 Can you create a separate issue for that? I can think of a few stats, such as `n_causal_paths`, `n_confounding_paths`, `n_exposure`, `n_outcomes`, something on mediators.

**Comment by Anushka-2406** [comment]
> @Anushka-2406 @DARHWOLF I wasn't thinking of causal statistics for this issue, as that would require the roles `exposures` and `outcomes` to be defined. @Anushka-2406 Can you create a separate issue for that? I can think of a few stats, such as `n_causal_paths`, `n_confounding_paths`, `n_exposure`, `n_outcomes`, something on mediators.

@ankurankan Sure, I'll add it.

File: pgmpy/tests/test_base/test_DAG.py [hunk 6]
```diff
@@ -718,6 +718,24 @@ def test_latents_without_role(self):
...
+        # D->G<-I (v-structure), G->L, I->S, isolated node J
+        from pgmpy.utils import get_example_model
+
+        model = get_example_model("sachs")
```
Comment (ankurankan) [inline pgmpy/tests/test_base/test_DAG.py]: `get_example_model` is deprecated. Please use `pgmpy.example_models.load_model` instead.

File: pgmpy/tests/test_base/test_DAG.py [hunk 7]
```diff
@@ -718,6 +718,24 @@ def test_latents_without_role(self):
         self.assertEqual(self.dag1.latents, set())
         self.assertEqual(set(self.dag1.get_role("latents")), set())
 
+    def test_get_stats(self):
+        # D->G<-I (v-structure), G->L, I->S, isolated node J
```
Comment (ankurankan) [inline pgmpy/tests/test_base/test_DAG.py]: I don't think this comment is relevant anymore?

File: pgmpy/tests/test_base/test_DAG.py [hunk 8]
```diff
@@ -718,6 +718,24 @@ def test_latents_without_role(self):
...
+        from pgmpy.utils import get_example_model
+
+        model = get_example_model("sachs")
+        dag = DAG(model.edges())
```
Comment (ankurankan) [inline pgmpy/tests/test_base/test_DAG.py]: Bayesian Network classes inherit DAG. You can directly call the `get_stats` method on `model`.

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by Anushka-2406** [comment]
@ankurankan @DARHWOLF 
I have made the changes and added n_latent_nodes in this PR. Once it gets merged, I will complete issue #2993.

**Review by ankurankan — CHANGES_REQUESTED** [review]
@Anushka-2406 I see that the `n_v_structures` in the tests for sachs model is 0. This is incorrect. You can see in the model here: https://www.bnlearn.com/bnrepository/discrete-small.html#sachs that there are 9 v-structures. Can you check why the value is incorrect?

**Comment by Anushka-2406** [comment]
> @Anushka-2406 I see that the `n_v_structures` in the tests for sachs model is 0. This is incorrect. You can see in the model here: https://www.bnlearn.com/bnrepository/discrete-small.html#sachs that there are 9 v-structures. Can you check why the value is incorrect?

Yeah I thought that too but when I found online there is a condition that both the parents should not be connected to each other. So that's why it's zero.

**Comment by ankurankan** [comment]
@Anushka-2406 You are right. I missed that all of them had direct edges between them.

**Review by ankurankan — APPROVED** [review]

**Review by DARHWOLF — APPROVED** [review]
